Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+  RsaPublicKey = (RSA_PUBLIC_KEY *) (AllocateZeroPool (RsaKeyStructLen));
+  
+  if (RsaPublicKey == NULL) {
+    DEBUG ((DEBUG_INFO, "RsaPublicKey allocation failure.\n"));
+    return NULL;
+  }
+
+  RsaPublicKey->Size = KeyHeader.KeyNumBits / 32;
+  RsaPublicKey->N0Inv = KeyHeader.N0Inv;
+  RsaPublicKey->N = (UINT32 *) (RsaPublicKey + 1);
+  RsaPublicKey->Rr = RsaPublicKey->N + RsaPublicKey->Size;
+
+  for (Index = 0; Index < RsaPublicKey->Size; Index++) {
+    RsaPublicKey->N[Index] = ((UINT32 *) N)[RsaPublicKey->Size - Index - 1];
+    RsaPublicKey->Rr[Index] = ((UINT32 *) Rr)[RsaPublicKey->Size - Index - 1];
+  }
+  
+  return RsaPublicKey;
+}
+
+
+/**
+  Verify a SHA256WithRSA PKCS#1 v1.5 signature against an expected
+  SHA256 hash.
+
+  @param Key         RSA public key
+  @param Signature   RSA signature
+  @param Sha256      SHA-256 digest of the content to verify
+
+  @return FALSE on failure, TRUE on success.
+ **/
+BOOLEAN
+RsaVerify (
+  UINT8        *Key,
+  UINTN        KeyNumBytes,
+  CONST UINT8  *Signature,
+  UINTN        SigNumBytes,
+  CONST UINT8  *Hash,
+  UINTN        HashNumBytes,
+  CONST UINT8  *Padding,
+  UINTN        PaddingNumBytes
+  )
+{
+  BOOLEAN         IsSuccess;  
+  RSA_PUBLIC_KEY  *ParsedKey;
+  UINT8           *WorkBuffer;
+  UINTN           *CalulatedPaddingNumBytes;
+
+  IsSuccess = FALSE;
+  ParsedKey = NULL;
+  WorkBuffer = NULL;
+
+  //
+  // Check input data
+  //
+  if (Key == NULL || Signature == NULL || Hash == NULL || Padding == NULL) {
+    DEBUG ((DEBUG_INFO, "Invalid input.\n"));
+    goto Exit;
+  }
+
+  //
+  // Parse key data
+  //
+  ParsedKey = RsaParseKeyData(Key, KeyNumBytes);
+  if (ParsedKey == NULL) {
+    DEBUG ((DEBUG_INFO, "Error parsing key.\n"));
+    goto Exit;
+  }
+
+  //
+  // Check signature length
+  //
+  if (SigNumBytes != (ParsedKey->Size * sizeof (UINT32))) {
+    DEBUG ((DEBUG_INFO, "Signature length does not match key length."));
+    goto Exit;
+  }
+
+  if (OcOverflowSubUN (SigNumBytes, HashNumBytes, CalulatedPaddingNumBytes)) {
+    DEBUG ((DEBUG_INFO, "Integer overflow while calculating real padding num bytes."));
+    goto Exit;
+  }
+
+  if (PaddingNumBytes != *CalulatedPaddingNumBytes) {
+    DEBUG ((DEBUG_INFO, "Padding length does not match hash and signature lengths."));
+    goto Exit;    
+  }
+
+  WorkBuffer = (UINT8 *) AllocateZeroPool (SigNumBytes);
+  if (WorkBuffer == NULL) {
+    DEBUG ((DEBUG_INFO, "WorkBuffer allocation failure.\n"));
+    goto Exit;    
+  }
+
+  //
+  // Copy input to local workspace
+  //
+  CopyMem (WorkBuffer, Signature, SigNumBytes);
+
+  //
+  // In-place exponentiation
+  //
+  ModPow (ParsedKey, WorkBuffer);
+
+  //
+  // Check padding bytes.
+  //
+  if (SecureCompareMem (WorkBuffer, Padding, PaddingNumBytes)) {
+    DEBUG ((DEBUG_INFO, "Padding check failed.\n"));
+    goto Exit;   
+  }
+
+  //
+  // Check the hash digest
+  //
+  if (SecureCompareMem (WorkBuffer + PaddingNumBytes, Hash, HashNumBytes)) {
```

Comment:
The comparison is done wrong. You compare and abort above, which effectively results in time difference.

Reply:
Solved. https://github.com/acidanthera/OcSupportPkg/commit/af65a5b7a61e8866962b180b27f836ca8e6befb3

Reply:
Not solved. Actually I missed the key issue here, in regards to comparison condition being written not properly, but what I was talking originally about is the time difference.

Comparing just the padding bits will result in faster reply than comparing hash and padding. What we should write here is basically:

```
volatile UINT32 Failure;

Failure = SecureCompareMem (…) != 0;
Failure += SecureCompareMem (…) != 0;
```

Reply:
This is obviously safe for out current return convention for SecureCompareMem and the few calls, but possibly we should conventionally use OR operations to not even have to think about wraparound issues and such?